M2/Macaulay2/d/gmp.d

```diff
@@ -1185,30 +1188,58 @@ export (x:CC) === (y:QQ) : bool := x.re === y && x.im === 0;
 export (x:QQ) === (y:CC) : bool := x === y.re && y.im === 0;
 
 export compare(x:CC,y:CC):int := (
+     if ( isinf(x.re) || isinf(y.re) || isinf(x.im) || isinf(y.im) ) then (
+       setflag0();
+       return 0;
+     );
      r := compare(x.re,y.re);
      if flagged() || r != 0 then r
      else compare(x.im,y.im));
 export compare(x:CC,y:RR):int := (
+     if ( isinf(x.re) || isinf(x.im) || isinf(y) ) then (
+       setflag0();
+       return 0;
+     );
      r := compare(x.re,y);
      if flagged() || r != 0 then r
      else compare0(x.im,0));
 export compare(x:RR,y:CC):int := (
+     if ( isinf(x) || isinf(y.re) || isinf(y.im) ) then (
+       setflag0();
+       return 0;
+     );
      r := compare(x,y.re);
      if flagged() || r != 0 then r
      else -compare0(y.im,0));
 export compare(x:CC,y:ZZ):int := (
+     if ( isinf(x.re) || isinf(x.im) ) then (
+       setflag0();
+       return 0;
+     );
      r := compare(x.re,y);
      if flagged() || r != 0 then r
      else compare0(x.im,0));
 export compare(x:ZZ,y:CC):int := (
+     if ( isinf(y.re) || isinf(y.im) ) then (
+       setflag0();
+       return 0;
+     );
      r := compare(x,y.re);
      if flagged() || r != 0 then r
      else -compare0(y.im,0));
 export compare(x:CC,y:QQ):int := (
+     if ( isinf(x.re) || isinf(x.im) ) then (
+       setflag0();
+       return 0;
+     );
      r := compare(x.re,y);
      if flagged() || r != 0 then r
      else compare0(x.im,0));
 export compare(x:QQ,y:CC):int := (
+     if ( isinf(y.re) || isinf(y.im) ) then (
+       setflag0();
```

Comment:
Yes, there is another place in the code that checks whether the flag has been set. Specifically, in actors3.d, the definition of `compare` on line 212 has "`if flagged() then incomparableE else`" in several branches.

I'm afraid I don't have a very clear memory of the specific purpose of setting this flag. I think the idea was to catch things like
``1. + 2.*ii ? 2. + (1/0.)*ii``
which should return `incomparable` (the second complex number is complex infinity) but without the flag, would just return `<` based on the comparison of real parts. To be pedantic there are two issues. One is that it needs to check if any imaginary parts are infinite before blindly comparing real parts. I guess it is a little redundant to also check if the real parts are infinite, as any infinities in the real parts would be caught and flagged when the real parts are compared. Ditto, some of the checking in other cases (e.g., line 1198, `compare(x:CC,y:RR)`, it would be sufficient to check if `x.im` is infinite); the extra checks could be stripped out but I think they do no harm. The second issue is that this level of compare can't pass `incomparable` up to the next level except in a GMP flag, as far as I can see. If there is another way, it could be used instead.